### ISSUE TITLE
Remove features deprecated when ksh93 was released

### ISSUE THREAD

**Issue by krader1961** [att/ast#344]
See the *src/cmd/ksh93/OBSOLETE* file for a list of features flagged as deprecated with the ksh93 release. We should give serious thought to removing them in the next major release. Even the backtick command substitution form should probably be removed so that there is just one way to do command substitutions. Some people are going to deem this too radical for consideration. I disagree. How long does a feature have to be deprecated before it can be removed? I think twenty-four years is long enough :smile:

**Comment by siteshwar** [comment]
One of the reasons `ksh` is still used today is because there are large number of scripts written in it. Deprecating things like backticks will only end up creating chaos.

**Comment by kdudka** [comment]
I think the answer is obvious: the ``` `...` ``` syntax can never ever be removed.  What would be worth to consider is making ksh interpret ``` `...` ``` and ```$(...)``` in the same way.  But even that is going to break some existing scripts I am afraid.  Similarly with the ```f() {...}``` syntax and ```function f() {...}``` syntax, which are each interpreted differently by ksh.  As far as compatibility matters, it would be IMO best to just analyze the current design and document it properly.

**Comment by krader1961** [comment]
> Deprecating things like backticks will only end up creating chaos.

That implies that ksh must remain compatible with the [POSIX 1003 standard](http://pubs.opengroup.org/onlinepubs/007908799/xcu/chap2.html) even in a major release. That standard for shells dates back to at least 1992. Which means that standard is at least twenty-five years old. Like many standards its main purpose was to codify long established behavior. Rather than define new and better behavior.

I am okay with formalizing that stance for this project but doing so means that ksh will never be competitive with general purpose languages like Python. Let alone other, newer, shells like [Fish](https://fishshell.com/) and [Elvish](https://elvish.io). It is not obvious how to reconcile statements from other individuals in other Github issues regarding making ksh competitive with language like Python in light of the restriction to remain POSIX 1003.2 compatible.

I was hoping to gently drop support for POSIX 1003 in a few key areas. Such as how variable expansion and word splitting interact. At least when the variable has been explicitly declared to be an array or map. But that will be impossible if we can never deviate from POSIX 1003.2 compliance in any way. In which case what is the purpose of this project? Is it solely to make it easier to fix bugs to a version of ksh that most distros adopted five years ago?

**Comment by siteshwar** [comment]
POSIX compatibility is a major feature of ksh. If we deviate from it, there is a big risk that users are going to reject our changes. Even in case of shells like fish, there are discussions around supporting POSIX syntax. It will be a mistake to break compatibility with POSIX standard.

**Comment by krader1961** [comment]
My `Even the backtick command substitution form should probably be removed` assertion was written tongue-in-cheek. I don't think we can realistically remove support for that feature without making a major release with an explicit declaration support for  backtick command substitution will be removed. My point is that too many discussions of this nature implicitly assume existing behavior can never be invalidated. Which means the project will die a slow  death due to an ability to evolve.

**Comment by siteshwar** [comment]
> My point is that too many discussions of this nature implicitly assume existing behavior can never be invalidated. 

I think we should take such discussions individually. I would like to hear more feedback from users when we attempt to make any change. This GitHub project still lacks visibility,  but I hope it will improve over time. 

> Which means the project will die a slow death due to an ability to evolve.

We can not survive long term without surviving short term. Any major breakage with existing ksh versions will guarantee a sudden death.

**Comment by dannyweldon** [comment]
I believe it is possible to make `ksh` comparable to `python` _without_ breaking `POSIX`.  FWIW, I have written up quite a few proposals for syntax enhancements to do this but have not shared them yet.  There are about a dozen and I'm not sure whether to create one large issue or separate issues for each one.  What about one issue as an overview and a separate issue for each item so they can be discussed?

Now, I'm not saying I have all the answers, or even the coding ability to do it, and there could be unforeseen flaws in my proposals, but I'm passionate about it because if it can be pulled off correctly, `ksh` would stand head and shoulders, no make that waist and up above the other shells, bringing shell programming up to modern standards while maintaining `POSIX` compatibility.

**Comment by siteshwar** [comment]
@dannyweldon I would suggest to create separate issues for each enhancement (unless enhancements are really close to each other).

**Comment by krader1961** [comment]
@dannyweldon, Yes, please open a separate issue for each enhancement you have in mind. Issues that cover more than a single, distinct, topic are problematic if for no other reason than it becomes hard to know when the issue has been resolved. If you have ideas on how to make `ksh` as safe to use as non-POSIX shells like [Elvish](https://github.com/elves/elvish) we definitely want to hear them.

I'm skeptical `ksh` can be turned into a language comparable to something like `python` while retaining POSIX compatibility but I'd love to be proven wrong. Note that if the idea is to add an option to disable certain undesirable POSIX behaviors that's fine but obviously means a huge amount of additional complexity because you're introducing behavior "personalities". Python mostly solves this by limiting the scope of a `import` or `from ... import` statement to the current module. Including, notably, `from __future__ import X` statements. What would the equivalent look like for ksh?

**Comment by dannyweldon** [comment]
Thanks @krader1961 and @siteshwar, I'll start cleaning up what I have got so far.  Look, I may have flaws in my designs but lets start the discussion anyway and see what we can come up with.

I doubt if we could do anything as advanced as that `python` example, but the closest I can think of would be to add an option, but implement something like `zsh`'s `setopt localoptions` to make it possible to scope options to functions or inside new types.  My ideas don't involve adding options though.

**Comment by kdudka** [comment]
On Wednesday, January 10, 2018 5:49:51 AM CET Kurtis Rader wrote:
> I'm skeptical `ksh` can be turned into a language comparable to something
> like `python` while retaining POSIX compatibility but I'd love to be proven
> wrong.

If your intention is to develop a modern programming language, why are you 
starting with (k)sh in the first place?

The language is neither efficient, nor intuitive, compared to other modern 
programming languages.  The key advantage of shell is portability.  If your 
goal is to develop a shell-like language that is not compatible with POSIX 
shell (or bash), it is not going to be useful for anything but programming 
exercises.

**Comment by krader1961** [comment]
We're in violent agreement, @kdudka. My reply was meant for the individuals (including he who shall not be named: j.e.l.l.m.d.) who have repeatedly argued that ksh is a great general purpose language. Which is why, they argue, it makes perfect sense (/eye-roll) to expose all of the libm functions. They need to show us how ksh can evolve such that it really can be used as a general purpose language that is competitive with something like Python or Ruby. I don't think it can be done. At least not without something like a `--no-posix` flag that radically changes how it behaves. And if we have to resort to such a tactic what would be the point? Why not just use one of the myriad mature languages that is likely already better suited to solving the problem?

**Comment by krader1961** [comment]
FWIW, The point of this issue is to spark discussion about features marked obsolete long ago we might be able to jettison. The backtick form of command substitution is probably not in that set of features since it is codified by POSIX 1003.2. The reason I mentioned it is to move the ["Overton window"](https://en.wikipedia.org/wiki/Overton_window) about ideas that are acceptable to tolerate. Especially with respect to  removing features that were either

a) deprecated decades ago, or

b) were added in the past decade but are incomplete and/or broken (e.g., `SHOPT_FIXEDARRY`) or not compatible with the goals of the project (e.g., faithful `bash` emulation via `SHOPT_BASH`).

With respect to backtick command substitution all I can say is that no one I know has used it in more than two decades. Whether we should still support it for the sake of being able to act as a faithful `/bin/sh` implementation from the 1980's, and thus execute very old scripts which might use that syntax, is an open question. The answer may very well be "yes, you idiot". But I think it is worth discussing given that we're trying to resurrect ksh from its dormancy and make it relevant again.

**Comment by krader1961** [comment]
Removing support for backticks to do command substitution might be unacceptable but what about support for the `FCEDIT` env var? That was mentioned in *src/cmd/ksh93/OBSOLETE* and is thus so old that anyone still using it should not be too surprised if we no longer recognize that env var.

The point of this issue is not whether any particular feature should be removed. It is whether support for any legacy, deprecated, feature can be removed.

**Comment by kdudka** [comment]
I think that our understanding of the marketplace for ksh fundamentally differs.  You want keep ksh relevant by removing questionable (unused, broken, obsolete, deprecated, legacy, whatever) features, thus make it better maintainable, extensible, etc.  I want to keep ksh relevant by preserving its compatibility with already existing ksh scripts, which I think is what the current user base asks for in the first place.

Even if you improve ksh such that it compiles in 20 seconds, runs 10 times faster, and supports python-like module loading, nobody is going to use it unless it stays sufficiently compatible with ksh93.  Most of the existing users would be lost by loosing the compatibility and new users of such a shell would be difficult to find.

I do not understand your comment about removing support of features instead of removing features.  There is no (guaranteed) support of ksh by upstream.  It is mainly Red Hat who has been providing support to existing users of ksh.  And it is obvious that the view of what is supported and what not currently differs between Red Hat and ksh upstream.

**Comment by hlangeveld** [comment]
On 01/09/2018 02:56, Kurtis Rader wrote:
> See the /src/cmd/ksh93/OBSOLETE/ file for a list of features flagged as deprecated with the ksh93 release. We should give serious 
> thought to removing them in the next major release. Even the backtick command substitution form should probably be removed so that 
> there is just one way to do command substitutions. Some people are going to deem this too radical for consideration. I disagree. How 
> long does a feature have to be deprecated before it can be removed? I think twenty-four years is long enough 😄
> 

;-)

I think that there is little value in removing features.

But.

It would be {useful,interesting} to add one or two features that will warn or throw a fatal error
during compilation for either:

1. deprecated syntax
2. non-posix syntax

Cheers,
Henk Langeveld

**Comment by krader1961** [comment]
@kdudka: Yes, it does appear we have fundamentally different views for the future of ksh. You seem to want to retain 100% compatibility with the version of `ksh` included in distros such as Fedora and RHEL. With the possible exception of fixing bugs. I say "possible exception" because I have seen instances where fixing a bug is itself considered unacceptable because it deviates from existing behavior that users have come to expect. Hence my use of the phrase "bug for bug compatible" in another issue from when I worked for Sequent Computer Systems. But that means we can't even change the build system because doing so may introduce differences in behavior. Let alone doing something like removing ksh's dependency on the AST VMALLOC subsystem.

My view is that if ksh has any hope of remaining relevant we have to break with the past. Which means updating the build system. Which means dropping dependencies on things like VMALLOC and SFIO which have not, and will never be, adopted by our industry. Which means restyling the code to a consistent style (whatever that style is) so that people who want to contribute improvements can more easily grok the code. It means removing the half-baked features that are undocumented and untested (e.g., `SHOPT_FIXEDARRAY`). It means removing features of dubious value that are undocumented (e.g., `SHOPT_AUDIT`). It especially means removing ill-advised features that should never have been implemented in the first place (e.g., `SHOPT_BASH`).

I'm willing to bet that 99% of those modernization steps will not be noticed by anyone using `ksh`.

Removing support for backtick command substitution isn't something we should do until we've made at least one major release since AT&T open sourced this code. That proposal was deliberately provocative to force people to think about what changes are acceptable in the near term. Also, if you look at other shells like [Fish](https://github.com/fish-shell/fish-shell/) which explicitly are not POSIX 1003.2 compatible you will see that they regularly get questions about supporting `$(...)` syntax. But I've never seen them get a question about supporting the older backtick form. So while we probably should not, at least in the immediate future, remove support for backtick syntax I do not consider doing so unreasonable as a future change.

Vendors like Red Hat can obviously continue to maintain their own forks based on branch `2016-01-10-beta` and try to fix bugs in those forks. I wonder how easy that has been to do given all the issues that have caused me to champion changing the build system, restyling the code, removing support for pre-ANSI C compilers, etc. in order to update the code base to something that is easier to understand and work with.

**Comment by krader1961** [comment]
> Even if you improve ksh such that it compiles in 20 seconds, runs 10 times faster... nobody is going to use it unless it stays sufficiently compatible with ksh93.

I agree with that statement. The question is what is "sufficiently compatible with ksh93". If we remove the `tgammal` math function is that insufficiently compatible with the last public ksh93 release? What about the behavior predicated on the `SHOPT_AUDIT` symbol? My first question is obviously a straw man question since only a handful of individuals will argue for continuing to support that math function despite being unable to provide evidence anyone has used them in a production ksh script. Removing the behavior predicated on the `SHOPT_AUDIT` is more difficult. But I think I can make the case the current implementation is does not achieve its goals and is not sufficiently documented.

**Comment by catull** [comment]
Could the mathematical functions be extracted into a "plugin" - i.e. shared library.
It would only be loaded if an environment variable were set, e.g. `KSH_MATH_ENABLED`.
Also, as soon as one of the math functions supported are invoked.

**Comment by kdudka** [comment]
On Friday, January 12, 2018 7:45:11 AM CET Kurtis Rader wrote:
> I'm willing to bet that 99% of those modernization steps will not be noticed
> by anyone using `ksh`.

I wish you were right.  Unfortunately, we have currently no data to support 
this statement.  The only reliable feedback we have is that a customer opens
a support case about something that broke after update.  At that time, it is 
usually already too late to do anything reasonable about it.  Red Hat simply 
cannot afford this try/fail approach.

If I could recommend anything, please release a new version of ksh consumable 
by Linux distros rather sooner than later.  This way you can get feedback on 
the recent changes before it is too late to revisit some of them.

Otherwise the development feels to me just too disconnected from end users
to be successful.

**Comment by dannyweldon** [comment]
Regarding the math functions, they are an integral part of `ksh93` now, (that ship has already left) so can't just be removed or deprecated on a whim.  Why should they be removed anyway, and how is that improving `ksh93` and making it more relevant?

And even if things are buggy still, that doesn't mean they should be removed?  Isn't your job as a maintainer to fix the bugs, not just to remove the code because you think no-one is using or would use it?  Including `SHOPT_FIXEDARRY`.  And why is the burden of proof whether people are using math functions on us anyway?  There was a lot of interest in the `ast-developers` list for math functions, which is why they are there.


To quote Michal Hlavinka, a previous `ksh93` maintainer from Red Hat from the `ast-developers` mailing list (not indexed on the web):

> We ship what version our userbase needs and they mostly need their scripts to be working. Some have scripts hundreds of thousands lines long and don't want to update those scripts (ksh sometimes changes behaviour with next release) because it would be huge effort and it would introduce bugs. I'm not sure about bug infested either. We track more complains about bugs in latest release than the patched version we have. (Talking about real complains, not theoretical list of what bug can anyone see by checking the changelog.) The bugs that our userbase reports are usually present in latest release too and we usually write/backport the fix for next rhel minor release. So, lets agree to disagree.
> ...
> Back to the topic. ksh in rhel 7 won't be updated any time soon. First, there is no stable release to update to. Second, it changes behaviour of `...` vs. $(...) and when it's supposed to be completed. We can't throw that at our users in released version. It would break their scripts and make them angry. So this will get in in next major update (rhel 8). If you want this to be backported, you have to contact the official support and it will be evaluated by product management.

So even 40-year-old shell syntax has to stay, which is why `ksh` has always just added new syntax to get around old limitations.  Eg: `[[ ]]`, `$( )`, `function` keyword

To @kdudka:
> If your intention is to develop a modern programming language, why are you starting with (k)sh in the first place? 

For the love of shell scripting (`ksh93`)!  And enhancing `ksh93` with modern syntax would be awesome!

I have been very busy so have not had time to review my proposals and would rather be doing that than replying to this thread.

**Comment by DavidMorano** [comment]
@catull wrote:
> Could the mathematical functions be extracted into a "plugin" - i.e. shared library.
> It would only be loaded if an environment variable were set, e.g. KSH_MATH_ENABLED.
> Also, as soon as one of the math functions supported are invoked.

First, a dynamically loaded math library is ONLY needed if one wants to (for whatever
reason) reduce the size of the default KSH binary on a system -- but does the inclusion 
of math functions really make the KSH binary so much bigger?

If it is indeed desired to reduce the size of a minimum KSH binary, I had a similar thought 
as @catull but slightly different.  The user could set something like:
KSH_LIBMATH=libkshmath.so (or whatever the user might have)
in the environment, and then when the first math function is invoked, the shared library
would be loaded (|dlopen(3dl)|).  Also, there should be both a default name and path to the
library as well as searching the LD_LIBRARY_PATH or DYLD_LIBRARY_PATH (or whatever),
so that math functions still work even in the absence of any environment variable. That is,
the math functions would always be enabled.  If the caller never calls any of them, then
no library would be loaded. If, for example KSH was installed on a system in:
/usr/local/bin/ksh
then the default associated KSH math library could be:
/usr/local/lib/libkshmath.so.1
et cetera, you all get the idea.

**Comment by siteshwar** [comment]
I agree with @kdudka. We should get our changes in hands of community as soon as possible. This is the main reason I spent time trying to get [Copr builds](https://copr.fedorainfracloud.org/coprs/g/ksh/latest/) working. I have managed to get builds working for RHEL (CentOS) along with Fedora. Hopefully we will be able to get more users to test these packages. If we end up removing any feature which is required by majority of users, we will have to revert back.

**Comment by krader1961** [comment]
> Regarding the math functions.... Why should they be removed anyway, and how is that improving ksh93 and making it more relevant?

@dannyweldon, We're ratholing a bit on the question of the math functions. The issue isn't math functions per se. There is absolutely nothing wrong with exposing 80% of the libm functions. What I find egregious was exposing the other 20% of those functions for no reason other than that they are provided by most libm/libc implementations and exposing them was trivial. I'm talking about functions that cannot reasonably, or probably even correctly, be used in a language like ksh. Functions like `tgammal`, `j0` and `errf`. I am not talking about things like the trigonometric functions or `ceil`. I've asked multiple times, both publicly and privately (in response to emails I received) for a single example of a production script that uses those functions. Crickets. I would fire anyone who wrote a ksh script that used those math functions. Ksh is not even remotely the right tool for a job that requires those functions.

Finally, note that the original code had a separate `iffe` feature detection test for every single libm function (which is why it was a major contributor to the slow build speed). So theoretically you could find yourself trying to run your script on a `ksh` implementation that did not provide the `tan()` function. Or any other math function. Ksh provides no fallback implementation if the system libm/libc libraries don't provide one of those functions. It will simply tell you "ksh: tan(1) : unknown function".

> If I could recommend anything, please release a new version of ksh consumable by Linux distros rather sooner than later.  This way you can get feedback on the recent changes before it is too late to revisit some of them.

Agreed, @kdudka. That's why we're working to stabilize the new build system and ensure ksh can be built and run on a representative set of distros. The fact we can't currently get a working ksh on OpenBSD 6.2 concerns me.

It is also clear from this discussion that a rebranding is going to be needed. Which was obliquely mentioned in #335. We'll need to make it easy for distros, including mechanisms like Homebrew for macOS, to provide both the legacy ksh93 side by side with this resurrected incarnation.

> So even 40-year-old shell syntax has to stay, 

Here I'm going to disagree, @dannyweldon. There are reasons why languages like Python do things like drop support for `print` as a language keyword and replace it with `print()` as a function when they went from python2 to python3. Yes, it was painful and many users choose to stay on python2. Nonetheless, such changes are necessary.

In any case, as I've said before, my suggestion we drop support for backtick command substitution wasn't really serious. While I hope no one has written a ksh script in the past decade using backtick command substitution I agree we can't remove support for it anytime soon. The question was to get people thinking about which incredibly old, deprecated, features we can drop when making the next major release. Better examples are things like `set -k` or the `FCEDIT` env var (both from *src/cmd/ksh93/OBSOLETE*).

> If we end up removing any feature which is required by majority of users, we will have to revert back.

I agree, @siteshwar. Although I might modify that statement by changing "majority" to "significant number" of users. I know "significant number" is an ill-defined phrase but I use it quite deliberately because there is a cost:benefit ratio to consider. We might retain a feature that is used by 0.1% of the ksh user base if that feature has a small source code footprint and is unlikely to cause problems over the long term. But we might jettison a feature used by 1.0% of the ksh user base if its source code footprint is large, it doesn't mesh well with the rest of ksh, is not compatible with the goals of the project (e.g., 100% compatibility with Bash), and/or is likely to be expensive to maintain.

**Comment by krader1961** [comment]
I can't help but mention an enhancement I'd like to see implemented to offset the perceived negatives of removing features deprecated decades ago. At some point I'll open an issue but the short version is ksh should allow marking the end of a block via a `end` keyword. Rather than require `fi` to end a `if` block, `esac` to end a `case` block, `done` for a `while` block, etcetera. The current syntax is not friendly. Imagine if C/C++ required something other than a close-brace, `}`, depending on the original flow of control statement.

**Comment by krader1961** [comment]
My prior comment noted that if ksh was built on a system that could not detect a math function like `tan()` it will emit this error if you try to use it:

> It will simply tell you "ksh: tan(1) : unknown function".

Which means no one can bitch about removing support for any math functions since ksh does not promise they will be available. So please explain to me why we cannot remove those functions  in light of the legacy project behavior. I agree that doing so might might break a script which uses those math functions. But by definition the legacy ksh could break that script.

Again, to be clear, I think ksh should continue to provide, unconditionally, support for basic math functions including trigonometric functions like `tan()`. If a platform's libm/libc does not support those functions the either ksh should not build on that platform or it should provide a fallback implementation.

**Comment by dannyweldon** [comment]
> I would fire anyone who wrote a ksh script that used those math functions. Ksh is not even remotely the right tool for a job that requires those functions.

Sorry, but when you keep making assertions like this, I think it just shows how out of touch you are with `ksh`.  And I do not have the time to respond to every argument/assertion you make that I disagree with or I would not get anything done.

I disagree that `fi`, `esac` and `done` are unfriendly as they indicate what type of block they are closing,  and I would be against what you are proposing.  An `end` keyword may have other unforeseen uses down the track, so wasting it on something like this would preclude that, and anyway, it just amounts to syntactic sugar.

> Which means no one can bitch about removing support for any math functions since ksh does not promise they will be available. So please explain to me why we cannot remove those functions in light of the legacy project behavior. I agree that doing so might might break a script which uses those math functions. But by definition the legacy ksh could break that script.

So, `ksh` is not going to implement the platform's math function library for them.  That's up to the OS maintainer.  If those functions are missing and someone wants to use them, they should simply log a support case with their vendor.

**Comment by hlangeveld** [comment]
On 01/13/2018 15:00, Danny Weldon wrote:
>     Which means no one can bitch about removing support for any math functions since ksh does not promise they will be available. So
>     please explain to me why we cannot remove those functions in light of the legacy project behavior. I agree that doing so might
>     might break a script which uses those math functions. But by definition the legacy ksh could break that script.
> 
> So, |ksh| is not going to implement the platform's math function library for them. That's up to the OS maintainer. If those 
> functions are missing and someone wants to use them, they should simply log a support case with their vendor.

+1

The math functions should be all-or-nothing.

However, I do understand Kurtis' argument about useless libm functions, and we could define
a standard set of libm functions that constitute a successful ksh math build?
On the other hand, that would mean we would still need to check for their presence
up front, which would slow down the build again...

On the third hand, we could postpone the checks with a post-build check of the math functions
from ksh itself.

Henk

**Comment by dannyweldon** [comment]
I don't think we need to provide `tgammal` or `tgammaf` versions of the math functions because they only take one type: `double`.

What if we added a phrase to the man age saying something like:

    subject to being available in the local C library implementation

or just:

    if available

**Comment by krader1961** [comment]
> I don't think we need to provide tgammal or tgammaf versions of the math functions because they only take one type: double.

The `ceil()` function also only takes one argument of type double. However, it is a math function that is somewhat easier to justify including as a standard builtin ksh math function since it is trivial to implement (if we had to provide a fallback implementation) and generally useful even in a shell script. So I don't understand your argument for not including `tgammal` and similar functions based on that criteria. Regardless, I am dropping my argument for removing libm math functions that should never have been exposed to ksh scripts. If and when we get a report of building ksh on a platform which does not provide those obscure functions, and which are not appropriate for a ksh script, we can revisit removing them.

**Comment by krader1961** [comment]
I'd like to shift the focus from discussion from the math functions back to what this issue is really about. Such as (from a prior comment):

> Better examples are things like `set -k` or the `FCEDIT` env var (both from *src/cmd/ksh93/OBSOLETE*).

I know this has implications for vendors like Red Hat and Suse who would like to to adopt and mainstream this code but can't afford to anger enterprise customers who might be relying on the obsolete features I think should be removed. If we proceed down this path we will probably need to do something similar to what the Python community did with the Python 2.x to 3.x transition. If we don't do that we will, at a minimum, need to make a release where those obsolete features are clearly deprecated followed by a major release no earlier than a year later that removes them.

The problem with maintaining support for features deprecated long ago is that it bloats the source code, makes it harder to reason about the programs behavior, and increases the maintenance costs.

**Comment by siteshwar** [comment]
One of our main goal for next release is to stay compatible with the last stable release, so I would not want to remove any user facing features (whether deprecated or not).

@krader1961 Unfortunately we don't have solid evidence that the features that you are proposing to remove are not used. As maintainers it would make our life simpler if we remove such features, but I don't want to do it at the risk of breaking existing scripts. I think we should close this issue.

**Comment by krader1961** [comment]
@siteshwar I have no objection to not removing any feature already present in ksh93u+ from the next release. Assuming we ever get this project into sufficiently good shape that a vendor would be willing to ship it. But the question still remains as to whether or not deprecated features can ever be removed. If the answer is no what is the point of deprecating them?
